bionic/executor.py

```diff
@@ -145,7 +145,28 @@ def __init__(self, queue):
         self._queue = queue
 
     def emit(self, record):
+        # TODO We send all log records to the main process for logging,
+        # even when logging is disabled for their log levels. This is
+        # expensive as we pay IPC cost for the log records and pay the
+        # cost of formating the log record, which can be costly for
+        # cases where users have a ton of disabled debug logs.
+        # We should investigate if we can either caching the log levels
+        # for each logger or communicate with main process to get the
+        # log level before we decide to send the logs.
         try:
+            # By adding the LogRecord to the queue, we send it to the
+            # process spawned by the multiprocessing manager. The
+            # communication between processes requires every argument
+            # to be pickleable using cloudpickle. Users can log
+            # unpickleable objects like clients or handlers and expect
+            # the logger to work. To avoid breaking logging, we eagerly
+            # format the log message the way logging does
+            # (check LogRecord.getMessage). This has a side effect of
+            # changing the template and arguments for the log in case
+            # users intercepts the LogRecord in their custom handlers.
+            if len(record.args) > 0:
+                record.msg = record.msg % record.args
```

Comment:
Should we make a copy of the record before mutating it? I'm wondering if there could be other handlers in this process that will also receive this record.

Reply:
Oh right, I guess users can add handlers to a logger inside the entity functions or we can use this outside Loky subprocesses too.